### ISSUE TITLE
az artifacts universal download not working behind proxy

### ISSUE THREAD

**Issue by muru03** [Azure/azure-devops-cli-extension#891]
## Describe the bug
I am working behind a corporate proxy. 
I am using below enviroment variables.
http_proxy: username:passoword@proxy.com:1234
https_proxy: username:passoword@proxy.com:1234
az login works fine. But the below command fails

**Command Name**
`az artifacts universal download
Extension Name: azure-devops. Version: 0.15.0.`

**Errors:**
```
HTTPSConnectionPool(host='08wvsblobprodsu6weus73.vsblob.vsassets.io', port=443): Max retries exceeded with url: /artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=ROwJZvi91FUWlgxNKLvGhncGvNt4pMAscin24AU8TpA%3D&spr=https&se=2019-11-19T12%3A28%3A02Z&sp=r&P1=1574166182&P2=11&P3=2&P4=nVC9zR0DzXvi2mMg1dV0dLTFHj9003iy92tc2KVTAk0%3d (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 407 Proxy Authentication Required',)))
Traceback (most recent call last):
pip-install-hb486cm7\urllib3\urllib3\connectionpool.py, ln 594, in urlopen
pip-install-hb486cm7\urllib3\urllib3\connectionpool.py, ln 805, in _prepare_proxy
pip-install-hb486cm7\urllib3\urllib3\connection.py, ln 308, in connect
...
```

## To Reproduce:
Steps to reproduce the behavior. Note that argument values have been redacted, as they may contain sensitive information.

- _Put any pre-requisite steps here..._
- `az artifacts universal download --organization {} --feed {} --name {} --version {} --path {}`

## Environment Summary
```
Windows-10-10.0.18362-SP0
Python 3.6.6
Shell: powershell.exe

azure-cli 2.0.76

Extensions:
azure-devops 0.15.0

```

**Comment by atbagga** [comment]
@muru03 Can you please share the debug logs for helping in troubleshooting. 
Run the same command with --debug.

**Comment by apluche** [comment]
@maru03 Is the system/IE proxy set?

**Comment by muru03** [comment]
> @muru03 Is the system/IE proxy set?
yes

**Comment by muru03** [comment]
> @muru03 Can you please share the debug logs for helping in troubleshooting.
> Run the same command with --debug.

Command arguments: ['artifacts', 'universal', 'download', '--organization', 'https://dev.azure.com/companyname', '--feed', 'companyname', '--name', 'packagename', '--version', '156.0.0', '--path', 'Packages', '--debug']
Event: Cli.PreExecute []
Event: CommandParser.OnGlobalArgumentsCreate [<function CLILogging.on_global_arguments at 0x03BE1A50>, <function OutputProducer.on_global_arguments at 0x03CAC738>, <function CLIQuery.on_global_arguments at 0x03D0F810>]
Event: CommandInvoker.OnPreCommandTableCreate []
Installed command modules ['acr', 'acs', 'advisor', 'ams', 'apim', 'appconfig', 'appservice', 'backup', 'batch', 'batchai', 'billing', 'botservice', 'cdn', 'cloud', 'cognitiveservices', 'configure', 'consumption', 'container', 'cosmosdb', 'deploymentmanager', 'dla', 'dls', 'dms', 'eventgrid', 'eventhubs', 'extension', 'feedback', 'find', 'hdinsight', 'interactive', 'iot', 'iotcentral', 'keyvault', 'kusto', 'lab', 'managedservices', 'maps', 'monitor', 'natgateway', 'netappfiles', 'network', 'policyinsights', 'privatedns', 'profile', 'rdbms', 'redis', 'relay', 'reservations', 'resource', 'role', 'search', 'security', 'servicebus', 'servicefabric', 'signalr', 'sql', 'sqlvm', 'storage', 'vm']
Loaded module 'acr' in 0.010 seconds.
Loaded module 'acs' in 0.024 seconds.
Loaded module 'advisor' in 0.004 seconds.
Event: CommandLoader.OnLoadCommandTable []
Loaded module 'ams' in 0.009 seconds.
Loaded module 'apim' in 0.005 seconds.
Loaded module 'appconfig' in 0.006 seconds.
Loaded module 'appservice' in 0.013 seconds.
Loaded module 'backup' in 0.006 seconds.
Event: CommandLoader.OnLoadCommandTable []
Loaded module 'batch' in 0.015 seconds.
Loaded module 'batchai' in 0.006 seconds.
Loaded module 'billing' in 0.005 seconds.
Loaded module 'botservice' in 0.006 seconds.
Event: CommandLoader.OnLoadCommandTable []
Loaded module 'cdn' in 0.007 seconds.
Loaded module 'cloud' in 0.004 seconds.
Loaded module 'cognitiveservices' in 0.004 seconds.
Loaded module 'configure' in 0.003 seconds.
Loaded module 'consumption' in 0.006 seconds.
Loaded module 'container' in 0.005 seconds.
Loaded module 'cosmosdb' in 0.012 seconds.
Loaded module 'deploymentmanager' in 0.004 seconds.
Loaded module 'dla' in 0.008 seconds.
Loaded module 'dls' in 0.005 seconds.
Loaded module 'dms' in 0.005 seconds.
Loaded module 'eventgrid' in 0.005 seconds.
Loaded module 'eventhubs' in 0.009 seconds.
Loaded module 'extension' in 0.004 seconds.
Loaded module 'feedback' in 0.002 seconds.
Loaded module 'find' in 0.004 seconds.
Loaded module 'hdinsight' in 0.005 seconds.
Loaded module 'interactive' in 0.001 seconds.
Loaded module 'iot' in 0.009 seconds.
Loaded module 'iotcentral' in 0.003 seconds.
Loaded module 'keyvault' in 0.009 seconds.
Loaded module 'kusto' in 0.004 seconds.
Loaded module 'lab' in 0.007 seconds.
Loaded module 'managedservices' in 0.004 seconds.
Loaded module 'maps' in 0.004 seconds.
Loaded module 'monitor' in 0.010 seconds.
Loaded module 'natgateway' in 0.004 seconds.
Event: CommandLoader.OnLoadCommandTable []
Loaded module 'netappfiles' in 0.007 seconds.
Loaded module 'network' in 0.041 seconds.
Loaded module 'policyinsights' in 0.005 seconds.
Loaded module 'privatedns' in 0.009 seconds.
Loaded module 'profile' in 0.004 seconds.
Loaded module 'rdbms' in 0.010 seconds.
Loaded module 'redis' in 0.004 seconds.
Loaded module 'relay' in 0.008 seconds.
Loaded module 'reservations' in 0.005 seconds.
Loaded module 'resource' in 0.012 seconds.
Loaded module 'role' in 0.010 seconds.
Loaded module 'search' in 0.006 seconds.
Loaded module 'security' in 0.005 seconds.
Loaded module 'servicebus' in 0.008 seconds.
Loaded module 'servicefabric' in 0.004 seconds.
Loaded module 'signalr' in 0.006 seconds.
Loaded module 'sql' in 0.012 seconds.
Loaded module 'sqlvm' in 0.005 seconds.
Event: CommandLoader.OnLoadCommandTable []
Loaded module 'storage' in 0.042 seconds.
Loaded module 'vm' in 0.021 seconds.
Loaded all modules in 0.480 seconds. (note: there's always an overhead with the first module loaded)
Extensions directory: 'C:\Users\{user}\.azure\cliextensions'
Found 1 extensions: ['azure-devops']
Extensions directory: 'C:\Users\{user}\.azure\cliextensions'
Extension compatibility result: is_compatible=True cli_core_version=2.0.76 min_required=2.0.69 max_required=None
Extensions directory: 'C:\Users\{user}\.azure\cliextensions'
Loaded extension 'azure-devops' in 0.051 seconds.
Event: CommandInvoker.OnPreCommandTableTruncate [<function AzCliLogging.init_command_file_logging at 0x03CEAD68>]
az_command_data_logger : command args: artifacts universal download --organization {} --feed {} --name {} --version {} --path {} --debug
metadata file logging enabled - writing logs to 'C:\Users\{user}\.azure\commands'.
Event: CommandInvoker.OnPreArgumentLoad [<function register_global_subscription_argument.<locals>.add_subscription_parameter at 0x03DFD3D8>]
Event: CommandInvoker.OnPostArgumentLoad []
Event: CommandInvoker.OnPostCommandTableCreate [<function register_ids_argument.<locals>.add_ids_arguments at 0x03E227C8>, <function register_cache_arguments.<locals>.add_cache_arguments at 0x03E228A0>]
Event: CommandInvoker.OnCommandTableLoaded []
Event: CommandInvoker.OnPreParseArgs [<function _documentdb_deprecate at 0x04A848E8>]
Event: CommandInvoker.OnPostParseArgs [<function OutputProducer.handle_output_argument at 0x03CAC780>, <function CLIQuery.handle_query_parameter at 0x03D0F858>, <function register_ids_argument.<locals>.parse_ids_arguments at 0x03E22858>, <function handler at 0x04B58978>, <function DevCommandsLoader.post_parse_args at 0x05087CD8>]
Extensions directory: 'C:\Users\{user}\.azure\cliextensions'
Extensions directory: 'C:\Users\{user}\.azure\cliextensions'
az_command_data_logger : extension name: azure-devops
az_command_data_logger : extension version: 0.15.0
Checking for a new ArtifactTool
Checking for ArtifactTool updates
Getting credential: azdevops-cli:https://dev.azure.com/companyname
Getting credential: azdevops-cli: default
Current cloud config:
AzureCloud
trying to get token (temp) for tenant 982035bb-0c38-4f2e-8ce0-49e49bd4ee04 and user first.last@companyname.com
attempting to read file C:\Users\{user}\.azure\accessTokens.json as utf-8-sig
adal-python : 3fe859d5-efa5-42e7-84b9-4bde5817fef9 - Authority:Performing instance discovery: ...
adal-python : 3fe859d5-efa5-42e7-84b9-4bde5817fef9 - Authority:Performing static instance discovery
adal-python : 3fe859d5-efa5-42e7-84b9-4bde5817fef9 - Authority:Authority validated via static instance discovery
adal-python : 3fe859d5-efa5-42e7-84b9-4bde5817fef9 - TokenRequest:Getting token from cache with refresh if necessary.
adal-python : 3fe859d5-efa5-42e7-84b9-4bde5817fef9 - CacheDriver:finding with query keys: {'_clientId': '...', 'userId': '...'}
adal-python : 3fe859d5-efa5-42e7-84b9-4bde5817fef9 - CacheDriver:Looking for potential cache entries: {'_clientId': '...', 'userId': '...'}
adal-python : 3fe859d5-efa5-42e7-84b9-4bde5817fef9 - CacheDriver:Found 3 potential entries.
adal-python : 3fe859d5-efa5-42e7-84b9-4bde5817fef9 - CacheDriver:Resource specific token found.
adal-python : 3fe859d5-efa5-42e7-84b9-4bde5817fef9 - CacheDriver:Returning token from cache lookup, AccessTokenId: b'6/gm/vztOo1ar4FjWF32jxuNHPu3XjPNJQ13nzCpuuc=', RefreshTokenId: b'miiK9kewK1zvXnJFQ+g+qWYgG+FXimut6ILetOVOK+E='
msrest.universal_http.requests : Configuring retry: max_retries=3, backoff_factor=0.8, max_backoff=90
Current cloud config:
AzureCloud
Azure devops telemetry enabled.
Logging telemetry to azure devops server.
msrest.universal_http.requests : Configuring retry: max_retries=3, backoff_factor=0.8, max_backoff=90
azext_devops.devops_sdk._file_cache : Loading cache file: C:\Users\{user}\.azure-devops\python-sdk\cache\options.json
azext_devops.devops_sdk._file_cache : attempting to read file C:\Users\{user}\.azure-devops\python-sdk\cache\options.json as utf-8-sig
azext_devops.devops_sdk.client : File cache hit for options on: https://dev.azure.com/companyname
msrest.universal_http.requests : Configuring retry: max_retries=3, backoff_factor=0.8, max_backoff=90
azext_devops.devops_sdk._file_cache : Loading cache file: C:\Users\{user}\.azure-devops\python-sdk\cache\resources.json
azext_devops.devops_sdk._file_cache : attempting to read file C:\Users\{user}\.azure-devops\python-sdk\cache\resources.json as utf-8-sig
azext_devops.devops_sdk.client : Route template: _apis/{area}/{resource}
azext_devops.devops_sdk.connection : File cache hit for resources on: https://dev.azure.com/companyname
azext_devops.devops_sdk.client : Api version '5.0-preview.1'
msrest.universal_http.requests : Configuring retry: max_retries=3, backoff_factor=0.8, max_backoff=90
Looking up current version of ArtifactTool...azext_devops.devops_sdk.client : POST https://dev.azure.com/companyname/_apis/CustomerIntelligence/Events

azext_devops.devops_sdk.client : Request content: [{'area': 'AzureDevopsCli', 'feature': 'artifacts', 'properties': {'Command': 'universal download', 'Args': 'feed name version path scope organization', 'ShellType': 'cmd', 'IsInteractive': 'True', 'OutputType': 'json', 'OrgPresentInCommand': True, 'ProjectPresentInCommand': False, 'RepoPresentInCommand': False}}]
msrest.universal_http : Configuring redirects: allow=True, max=30
msrest.universal_http : Configuring request: timeout=100, verify=True, cert=None
msrest.universal_http : Configuring proxies: ''
msrest.universal_http : Evaluate proxies against ENV settings: True
urllib3.connectionpool : Starting new HTTPS connection (1): dev.azure.com:443
azext_devops.devops_sdk.client : File cache hit for options on: https://companyname.vsblob.visualstudio.com
azext_devops.devops_sdk.client : Route template: _apis/{area}/{toolName}/{resource}
azext_devops.devops_sdk.client : Api version '5.0-preview'
azext_devops.devops_sdk.client : GET https://companyname.vsblob.visualstudio.com/_apis/clienttools/ArtifactTool/release?osName=Windows&arch=AMD64
azext_devops.devops_sdk.client : Request content: None
msrest.universal_http : Configuring redirects: allow=True, max=30
msrest.universal_http : Configuring request: timeout=100, verify=True, cert=None
msrest.universal_http : Configuring proxies: ''
msrest.universal_http : Evaluate proxies against ENV settings: True
urllib3.connectionpool : Starting new HTTPS connection (1): companyname.vsblob.visualstudio.com:443
urllib3.connectionpool : https://dev.azure.com:443 "POST /companyname/_apis/CustomerIntelligence/Events HTTP/1.1" 204 0
urllib3.connectionpool : https://companyname.vsblob.visualstudio.com:443 "GET /_apis/clienttools/ArtifactTool/release?osName=Windows&arch=AMD64 HTTP/1.1" 200 446
azext_devops.devops_sdk.client : Response content: b'{"name":"ArtifactTool","rid":"win10-x64","version":"0.2.128","uri":"https://08wvsblobprodsu6weus73.vsblob.vsassets.io/artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=dMmcMPr5M8npY39DzGMvZ1ugMGCJMU49wJsxNab7xoE%3D&spr=https&se=2019-11-26T06%3A21%3A03Z&sp=r&P1=1574748963&P2=11&P3=2&P4=KO0e%2bnrAOaWTMkawUQqTRCH1m4sZa1WK4ZLMwRNNkbw%3d"}'
Checking if we already have ArtifactTool release 'ArtifactTool_win10-x64_0.2.128'
Updating to ArtifactTool release ArtifactTool_win10-x64_0.2.128 since it doesn't exist at C:\Users\{user}\.azure\azuredevops\cli\tools\artifacttool\ArtifactTool_win10-x64_0.2.128
Downloading ArtifactTool from https://08wvsblobprodsu6weus73.vsblob.vsassets.io/artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=dMmcMPr5M8npY39DzGMvZ1ugMGCJMU49wJsxNab7xoE%3D&spr=https&se=2019-11-26T06%3A21%3A03Z&sp=r&P1=1574748963&P2=11&P3=2&P4=KO0e%2bnrAOaWTMkawUQqTRCH1m4sZa1WK4ZLMwRNNkbw%3d
urllib3.connectionpool : Starting new HTTPS connection (1): 08wvsblobprodsu6weus73.vsblob.vsassets.io:443
handling generic error
cli.azure.cli.core.util : The command failed with an unexpected error. Here is the traceback:

The command failed with an unexpected error. Here is the traceback:

cli.azure.cli.core.util : HTTPSConnectionPool(host='08wvsblobprodsu6weus73.vsblob.vsassets.io', port=443): Max retries exceeded with url: /artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=dMmcMPr5M8npY39DzGMvZ1ugMGCJMU49wJsxNab7xoE%3D&spr=https&se=2019-11-26T06%3A21%3A03Z&sp=r&P1=1574748963&P2=11&P3=2&P4=KO0e%2bnrAOaWTMkawUQqTRCH1m4sZa1WK4ZLMwRNNkbw%3d (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 407 Proxy Authentication Required',)))
Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-hb486cm7\urllib3\urllib3\connectionpool.py", line 594, in urlopen
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-hb486cm7\urllib3\urllib3\connectionpool.py", line 805, in _prepare_proxy
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-hb486cm7\urllib3\urllib3\connection.py", line 308, in connect
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\http\client.py", line 919, in _tunnel
    message.strip()))
OSError: Tunnel connection failed: 407 Proxy Authentication Required

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\adapters.py", line 449, in send
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-hb486cm7\urllib3\urllib3\connectionpool.py", line 638, in urlopen
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-hb486cm7\urllib3\urllib3\util\retry.py", line 399, in increment
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='08wvsblobprodsu6weus73.vsblob.vsassets.io', port=443): Max retries exceeded with url: /artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=dMmcMPr5M8npY39DzGMvZ1ugMGCJMU49wJsxNab7xoE%3D&spr=https&se=2019-11-26T06%3A21%3A03Z&sp=r&P1=1574748963&P2=11&P3=2&P4=KO0e%2bnrAOaWTMkawUQqTRCH1m4sZa1WK4ZLMwRNNkbw%3d (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 407 Proxy Authentication Required',)))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\knack\knack\cli.py", line 206, in invoke
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\commands\__init__.py", line 603, in execute
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\commands\__init__.py", line 661, in _run_jobs_serially
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\commands\__init__.py", line 652, in _run_job
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\exception_handler.py", line 31, in azure_devops_exception_handler
    reraise(*sys.exc_info())
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\six\six.py", line 693, in reraise
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\commands\__init__.py", line 631, in _run_job
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\commands\__init__.py", line 306, in __call__
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\__init__.py", line 485, in default_command_handler
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\artifacts\universal.py", line 98, in download_package
    return artifact_tool.download_universal(organization, project, feed, name, version, path, file_filter)
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool.py", line 43, in download_universal
    return self.run_artifacttool(organization, args, "Downloading")
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool.py", line 58, in run_artifacttool
    artifacttool_dir = self._artifacttool_updater.get_latest_artifacttool(organization)
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool_updater.py", line 42, in get_latest_artifacttool
    artifacttool_binary_path = self._get_artifacttool(organization)
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool_updater.py", line 73, in _get_artifacttool
    _update_artifacttool(release_uri, release_id)
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool_updater.py", line 96, in _update_artifacttool
    response = requests.get(uri, stream=True)
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\api.py", line 75, in get
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\api.py", line 60, in request
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\sessions.py", line 533, in request
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\sessions.py", line 646, in send
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\adapters.py", line 510, in send
requests.exceptions.ProxyError: HTTPSConnectionPool(host='08wvsblobprodsu6weus73.vsblob.vsassets.io', port=443): Max retries exceeded with url: /artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=dMmcMPr5M8npY39DzGMvZ1ugMGCJMU49wJsxNab7xoE%3D&spr=https&se=2019-11-26T06%3A21%3A03Z&sp=r&P1=1574748963&P2=11&P3=2&P4=KO0e%2bnrAOaWTMkawUQqTRCH1m4sZa1WK4ZLMwRNNkbw%3d (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 407 Proxy Authentication Required',)))
HTTPSConnectionPool(host='08wvsblobprodsu6weus73.vsblob.vsassets.io', port=443): Max retries exceeded with url: /artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=dMmcMPr5M8npY39DzGMvZ1ugMGCJMU49wJsxNab7xoE%3D&spr=https&se=2019-11-26T06%3A21%3A03Z&sp=r&P1=1574748963&P2=11&P3=2&P4=KO0e%2bnrAOaWTMkawUQqTRCH1m4sZa1WK4ZLMwRNNkbw%3d (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 407 Proxy Authentication Required',)))
Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-hb486cm7\urllib3\urllib3\connectionpool.py", line 594, in urlopen
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-hb486cm7\urllib3\urllib3\connectionpool.py", line 805, in _prepare_proxy
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-hb486cm7\urllib3\urllib3\connection.py", line 308, in connect
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\http\client.py", line 919, in _tunnel
    message.strip()))
OSError: Tunnel connection failed: 407 Proxy Authentication Required

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\adapters.py", line 449, in send
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-hb486cm7\urllib3\urllib3\connectionpool.py", line 638, in urlopen
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-hb486cm7\urllib3\urllib3\util\retry.py", line 399, in increment
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='08wvsblobprodsu6weus73.vsblob.vsassets.io', port=443): Max retries exceeded with url: /artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=dMmcMPr5M8npY39DzGMvZ1ugMGCJMU49wJsxNab7xoE%3D&spr=https&se=2019-11-26T06%3A21%3A03Z&sp=r&P1=1574748963&P2=11&P3=2&P4=KO0e%2bnrAOaWTMkawUQqTRCH1m4sZa1WK4ZLMwRNNkbw%3d (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 407 Proxy Authentication Required',)))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\knack\knack\cli.py", line 206, in invoke
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\commands\__init__.py", line 603, in execute
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\commands\__init__.py", line 661, in _run_jobs_serially
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\commands\__init__.py", line 652, in _run_job
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\exception_handler.py", line 31, in azure_devops_exception_handler
    reraise(*sys.exc_info())
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\six\six.py", line 693, in reraise
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\commands\__init__.py", line 631, in _run_job
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\commands\__init__.py", line 306, in __call__
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\azure-cli-core\azure\cli\core\__init__.py", line 485, in default_command_handler
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\artifacts\universal.py", line 98, in download_package
    return artifact_tool.download_universal(organization, project, feed, name, version, path, file_filter)
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool.py", line 43, in download_universal
    return self.run_artifacttool(organization, args, "Downloading")
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool.py", line 58, in run_artifacttool
    artifacttool_dir = self._artifacttool_updater.get_latest_artifacttool(organization)
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool_updater.py", line 42, in get_latest_artifacttool
    artifacttool_binary_path = self._get_artifacttool(organization)
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool_updater.py", line 73, in _get_artifacttool
    _update_artifacttool(release_uri, release_id)
  File "C:\Users\{user}\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool_updater.py", line 96, in _update_artifacttool
    response = requests.get(uri, stream=True)
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\api.py", line 75, in get
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\api.py", line 60, in request
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\sessions.py", line 533, in request
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\sessions.py", line 646, in send
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-umfcm_n2\requests\requests\adapters.py", line 510, in send
requests.exceptions.ProxyError: HTTPSConnectionPool(host='08wvsblobprodsu6weus73.vsblob.vsassets.io', port=443): Max retries exceeded with url: /artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=dMmcMPr5M8npY39DzGMvZ1ugMGCJMU49wJsxNab7xoE%3D&spr=https&se=2019-11-26T06%3A21%3A03Z&sp=r&P1=1574748963&P2=11&P3=2&P4=KO0e%2bnrAOaWTMkawUQqTRCH1m4sZa1WK4ZLMwRNNkbw%3d (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 407 Proxy Authentication Required',)))

**Comment by apluche** [comment]
Hi @muru03

A variety of things can be happening during this call so I would like to ask you to go through several steps to help us determine the root of the problem. 

**1) Use ArtifactTool directly**
Please follow the following link after filling out the org name and the os-name (os name should be windows or linux): https://vsblob.dev.azure.com/<organization-name>/_apis/clienttools/artifacttool/release?osName=<OS-name>&arch=x86_64 . From this request, you will receive a Json structure resembling this:

{
  "name": "artifacttool",
  "rid": "win10-x64",
  "version": "0.2.104",
  "uri": "https://03nvsblobprodwcus014.vsblob.vsassets.io/artifacttool/...."
}

Please download ArtifactTool from the returned URI and extract the tool from the zip onto your machine. ArtifactTool only supports PAT authentication. Documentation for how to generate a PAT can be find here: https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=preview-page . The PAT scope should contain Packaging permissions (read permissions needed for download, read&write needed for download and publish). The PAT will need to be passed as an environment variable. You will need to pass the environment variable name to a --patvar parameter. The rest of the parameters should be familiar and can be found in the Connect to Feed dialogue.

Ex: 
$env:ArtifactToolPat="<pat-goes-here>"
.\artifacttool.exe universal download --service <service url> --patvar ArtifactToolPat --feed <feedName> --package-name <packageName> --package-version <version> --path <path to download to>
.\artifacttool.exe universal publish --service <service url> --patvar ArtifactToolPat --feed <feedName> --package-name <packageName> --package-version <version> --path <path to file to publish> 

If this works, this means that you can access and use ArtifactTool correctly and it isn't a permissions/local configurations issue. This can be used as a workaround if the following steps don't work while we continue to investigate.

**2) Use the CLI but have the CLI run a local version of the ArtifactTool**
You will need to set an environment variable leading to the folder containing the ArtifactTool you downloaded in the previous step. The path should be to the folder containing the executable.

$env:AZURE_DEVOPS_EXT_ARTIFACTTOOL_OVERRIDE_PATH= "<path>\artifacttool\ArtifactTool_win10-x64_0.2.128"

You should be able to use the CLI normally now, the only difference is that it will not attempt to auto update to a newer version of ArtifactTool. Please make sure to have --debug enabled.

**3) Use the CLI with automatic download**

Remove the environment variable from the previous step and use normal CLI command.

Please let me know which step starts causing errors; any logs you get would be appreciated.

**Comment by muru03** [comment]
> Hi @muru03
> 
> A variety of things can be happening during this call so I would like to ask you to go through several steps to help us determine the root of the problem.
> 
> **1) Use ArtifactTool directly**
> Please follow the following link after filling out the org name and the os-name (os name should be windows or linux): https://vsblob.dev.azure.com/_apis/clienttools/artifacttool/release?osName=&arch=x86_64 . From this request, you will receive a Json structure resembling this:
> 
> {
> "name": "artifacttool",
> "rid": "win10-x64",
> "version": "0.2.104",
> "uri": "https://03nvsblobprodwcus014.vsblob.vsassets.io/artifacttool/...."
> }
> 
> Please download ArtifactTool from the returned URI and extract the tool from the zip onto your machine. ArtifactTool currently only supports PAT authentication. Documentation for how to generate a PAT can be find here: https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=preview-page . The PAT scope should contain Packaging permissions (read permissions needed for download, read&write needed for download and publish). The PAT will need to be passed as an environment variable. You will need to pass the environment variable name to a --patvar parameter. The rest of the parameters should be familiar and can be found in the Connect to Feed dialogue.
> 
> Ex:
> $env:ArtifactToolPat=""
> .\artifacttool.exe universal download --service --patvar ArtifactToolPat --feed --package-name --package-version --path
> .\artifacttool.exe universal publish --service --patvar ArtifactToolPat --feed --package-name --package-version --path
> 
> If this works, this means that you can access and use ArtifactTool correctly and it isn't a permissions/local configurations issue. This can be used as a workaround if the following steps don't work while we continue to investigate.
> 
> **2) Use the CLI but have the CLI run a local version of the ArtifactTool**
> You will need to set an environment variable leading to the folder containing the ArtifactTool you downloaded in the previous step. The path should be to the folder containing the executable.
> 
> Ex.
> $env:AZURE_DEVOPS_EXT_ARTIFACTTOOL_OVERRIDE_PATH= "\artifacttool\ArtifactTool_win10-x64_0.2.128"
> 
> You should be able to use the CLI normally now, the only difference is that it will not attempt to auto update to a newer version of ArtifactTool. Please make sure to have --debug enabled.
> 
> **3) Use the CLI with automatic download**
> 
> Remove the environment variable from the previous step and use normal CLI command.
> 
> Please let me know which step starts causing errors; any logs you get would be appreciated.

While hitting the url (https://vsblob.dev.azure.com/{company name}/_apis/clienttools/artifacttool/release?windows=&arch=x86_64) I am getting the below error.

{"$id":"1","innerException":null,"message":"No release could be found based on the provided information.","typeName":"Microsoft.VisualStudio.Services.BlobStore.WebApi.Exceptions.ClientToolNotFoundException, Microsoft.VisualStudio.Services.BlobStore.WebApi","typeKey":"ClientToolNotFoundException","errorCode":0,"eventId":3000}

**Comment by apluche** [comment]
@muru03 It appears the initial link was not pasted over properly. I apologize for the mistake. Here is the link: https://vsblob.dev.azure.com/<organization-name>/_apis/clienttools/artifacttool/release?osName=<OS-name>&arch=x86_64 

It looks like the link you are hitting is malformed and should have osName=windows.

**Comment by israelvaldez** [comment]
The second workaround worked for me, once you redirect the CLI to the download location you can get your packages.

The help for the tool does not show anything when you poke it, it always returns the same message, is there any special argument that I need to send?, I had to decompile the artifacttool.dll to find the class CommandBase and see what the arguments were:

artifacttool.exe universal download --service "https://dev.azure.com/<Your DevOps Org>/" --patvar ArtifactToolPat --feed "<Your feed>" --package-name "<Your Package Name>" --package-version "<Your Package Version>" --path "<Path to save files>"

This is what you get when you call --help, it seems that the commands are not wired up correctly to show up in the help for some reason.
C:\_b\d>artifacttool.exe command --help
Usage: artifacttool [arguments] [options]

Arguments:
  category      The category to which the following command belongs.
  command       The command under the given category to be invoked.

Options:
  -?|-h|--help  Show help information

I could never get the first workaround to work I always got the following issue even though I was setting up the environment variable as expected:
`{
"@t":"2019-12-11T15:24:35.8644432Z",
"@m":"[GetDedupManifestArtifactClientAsync] Try 1/5, non-retryable exception caught. Throwing. Details:\r\nNo LastRequestResponse on exception VssUnauthorizedException: TF400813: Resource not available for anonymous access. Client authentication required.",
"@i":"503505fd",
"@l":"Warning",
"@x":"Microsoft.VisualStudio.Services.Common.VssUnauthorizedException: TF400813: Resource not available for anonymous access. Client authentication required.\r\n at Microsoft.VisualStudio.Services.Common.VssHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.Common.VssHttpRetryMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\r\n at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)\r\n at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync(HttpRequestMessage message, HttpCompletionOption completionOption, Object userState, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync[T](HttpRequestMessage message, Object userState, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.Location.Client.LocationHttpClient.GetConnectionDataAsync(ConnectOptions connectOptions, Int64 lastChangeId, CancellationToken cancellationToken, Object userState)\r\n at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.GetConnectionDataAsync(ConnectOptions connectOptions, Int32 lastChangeId, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.ConnectAsync(ConnectOptions connectOptions, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.Location.LocationService.CreateDataProviderAsync(String location, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.Location.LocationService.ResolveLocationDataAsync(Guid locationAreaIdentifier, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.Location.LocationService.GetLocationDataAsync(Guid locationAreaIdentifier, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientInstanceAsync(Type managedType, Guid serviceIdentifier, CancellationToken cancellationToken, VssHttpRequestSettings settings, DelegatingHandler[] handlers)\r\n at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientServiceImplAsync(Type requestedType, Guid serviceIdentifier, Func`4 getInstanceAsync, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientAsync[T](CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.Content.Common.AsyncHttpRetryHelper`1.InvokeAsync(CancellationToken cancellationToken)",
"SourceContext":"ArtifactTool.Commands.UPackDownloadCommand",
"UtcTimestamp":"2019-12-11 15:24:35.864Z"
}`
Then this one after:
`{
"@t":"2019-12-11T15:24:35.9204996Z",
"@m":"Encountered an unexpected error.",
"@i":"b8be900d",
"@l":"Error",
"@x":"Microsoft.VisualStudio.Services.Common.VssUnauthorizedException: TF400813: Resource not available for anonymous access. Client authentication required.\r\n at Microsoft.VisualStudio.Services.Common.VssHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.Common.VssHttpRetryMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\r\n at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)\r\n at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync(HttpRequestMessage message, HttpCompletionOption completionOption, Object userState, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync[T](HttpRequestMessage message, Object userState, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.Location.Client.LocationHttpClient.GetConnectionDataAsync(ConnectOptions connectOptions, Int64 lastChangeId, CancellationToken cancellationToken, Object userState)\r\n at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.GetConnectionDataAsync(ConnectOptions connectOptions, Int32 lastChangeId, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.ConnectAsync(ConnectOptions connectOptions, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.Location.LocationService.CreateDataProviderAsync(String location, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.Location.LocationService.ResolveLocationDataAsync(Guid locationAreaIdentifier, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.Location.LocationService.GetLocationDataAsync(Guid locationAreaIdentifier, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientInstanceAsync(Type managedType, Guid serviceIdentifier, CancellationToken cancellationToken, VssHttpRequestSettings settings, DelegatingHandler[] handlers)\r\n at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientServiceImplAsync(Type requestedType, Guid serviceIdentifier, Func`4 getInstanceAsync, CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientAsync[T](CancellationToken cancellationToken)\r\n at Microsoft.VisualStudio.Services.Content.Common.AsyncHttpRetryHelper`1.InvokeAsync(CancellationToken cancellationToken)\r\n at ArtifactTool.DedupManifestArtifactClientProvider.GetDedupManifestArtifactClientAsync(String serviceUrl, String patVar, ILogger commandBaseLogger, IAppTraceSource tracer, CancellationToken cancellationToken) in d:\\a\\1\\s\\src\\ArtifactTool\\Providers\\DedupManifestArtifactClient\\DedupManifestArtifactClientProvider.cs:line 42\r\n at ArtifactTool.Commands.UPackDownloadCommand.ExecuteAsync() in d:\\a\\1\\s\\src\\ArtifactTool\\Commands\\UPack\\UPackDownloadCommand.cs:line 52\r\n at ArtifactTool.Commands.CommandBase.OnExecuteAsync() in d:\\a\\1\\s\\src\\ArtifactTool\\Commands\\CommandBase.cs:line 100\r\n at McMaster.Extensions.CommandLineUtils.Conventions.ExecuteMethodConvention.InvokeAsync(MethodInfo method, Object instance, Object[] arguments) in C:\\projects\\commandlineutils\\src\\CommandLineUtils\\Conventions\\ExecuteMethodConvention.cs:line 77\r\n at McMaster.Extensions.CommandLineUtils.Conventions.ExecuteMethodConvention.OnExecute(ConventionContext context) in C:\\projects\\commandlineutils\\src\\CommandLineUtils\\Conventions\\ExecuteMethodConvention.cs:line 62\r\n at McMaster.Extensions.CommandLineUtils.Conventions.ExecuteMethodConvention.<>c__DisplayClass0_0.<<Apply>b__0>d.MoveNext() in C:\\projects\\commandlineutils\\src\\CommandLineUtils\\Conventions\\ExecuteMethodConvention.cs:line 25\r\n--- End of stack trace from previous location where exception was thrown ---\r\n at McMaster.Extensions.CommandLineUtils.CommandLineApplication.<>c__DisplayClass126_0.<OnExecute>b__0() in C:\\projects\\commandlineutils\\src\\CommandLineUtils\\CommandLineApplication.cs:line 505\r\n at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args) in C:\\projects\\commandlineutils\\src\\CommandLineUtils\\CommandLineApplication.cs:line 611\r\n at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute[TApp](CommandLineContext context) in C:\\projects\\commandlineutils\\src\\CommandLineUtils\\CommandLineApplication.Execute.cs:line 57\r\n at McMaster.Extensions.CommandLineUtils.CommandLineApplication.ExecuteAsync[TApp](CommandLineContext context) in C:\\projects\\commandlineutils\\src\\CommandLineUtils\\CommandLineApplication.Execute.cs:line 145\r\n at McMaster.Extensions.CommandLineUtils.CommandLineApplication.ExecuteAsync[TApp](IConsole console, String[] args) in C:\\projects\\commandlineutils\\src\\CommandLineUtils\\CommandLineApplication.Execute.cs:line 130\r\n at McMaster.Extensions.CommandLineUtils.CommandLineApplication.ExecuteAsync[TApp](String[] args) in C:\\projects\\commandlineutils\\src\\CommandLineUtils\\CommandLineApplication.Execute.cs:line 112",
"SourceContext":"ArtifactTool.Program",
"UtcTimestamp":"2019-12-11 15:24:35.920Z"
}`

**Comment by apluche** [comment]
Hi @israelvaldez, 

I'm glad to hear that the second workaround worked for you. The first workaround is not really documented since it is better to use the artifacttool from the CLI rather than use it alone. Since you were able to just get the second one to work, please use the CLI with the specific artifacttool set. It will serve you better than using the tool directly. 

The help on the tool itself isn't very well done since it isn't advised to use it directly. Unfortunately, you didn't need to dig into the tool to find the command. I included the arguments that you would need to run the command in artifacttool in my previous comments but it looks like it also fell to formatting errors. I'm copying it here and forcing the right formatting:
**.\artifacttool.exe universal download --service {<service url>} --patvar ArtifactToolPat --feed {<feedName>} --package-name {<packageName>} --package-version {<version>} --path {<path to download to>}**

As for your stack trace, it looks like the ArtifactToolPat that you set didn't actually get set. I ran into this problem when doing it myself. Copy and pasting the PAT doesn't work well on the command line so you will need to make sure that the token gets correctly copied over. That being said, since the second workaround works for you, please use that one.

**Comment by apluche** [comment]
@muru03, did the workaround unblock you and are you able to download packages?

@muru03 , @israelvaldez Could you remove the environment variable (if it is still set) to see if you are still running into this issue? If so, please send the logs you receive.

**Comment by israelvaldez** [comment]
I get a max retries received error:
`The command failed with an unexpected error. Here is the traceback:

HTTPSConnectionPool(host='0t3vsblobprodcus362.vsblob.vsassets.io', port=443): Max retries exceeded with url: /artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=EWrX6QfaHgakt2Wx2sORhYViPB7P0jNWWbeu3yGP3Dg%3D&spr=https&se=2020-01-16T19%3A17%3A49Z&sp=r&P1=1579201969&P2=11&P3=2&P4=lKDYcrhMcfeD75HvPn3Y19rU94ccBKYN83qNwaa6qHs%3d (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')],)",),))
Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-2bnqpf73\urllib3\urllib3\contrib\pyopenssl.py", line 456, in wrap_socket
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\pyOpenSSL\OpenSSL\SSL.py", line 1915, in do_handshake
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\pyOpenSSL\OpenSSL\SSL.py", line 1647, in _raise_ssl_error
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\pyOpenSSL\OpenSSL\_util.py", line 54, in exception_from_error_queue
OpenSSL.SSL.Error: [('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-2bnqpf73\urllib3\urllib3\connectionpool.py", line 600, in urlopen
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-2bnqpf73\urllib3\urllib3\connectionpool.py", line 343, in _make_request
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-2bnqpf73\urllib3\urllib3\connectionpool.py", line 839, in _validate_conn
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-2bnqpf73\urllib3\urllib3\connection.py", line 344, in connect
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-2bnqpf73\urllib3\urllib3\util\ssl_.py", line 347, in ssl_wrap_socket
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-2bnqpf73\urllib3\urllib3\contrib\pyopenssl.py", line 462, in wrap_socket
ssl.SSLError: ("bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')],)",)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\requests\requests\adapters.py", line 449, in send
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-2bnqpf73\urllib3\urllib3\connectionpool.py", line 638, in urlopen
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-2bnqpf73\urllib3\urllib3\util\retry.py", line 399, in increment
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='0t3vsblobprodcus362.vsblob.vsassets.io', port=443): Max retries exceeded with url: /artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=EWrX6QfaHgakt2Wx2sORhYViPB7P0jNWWbeu3yGP3Dg%3D&spr=https&se=2020-01-16T19%3A17%3A49Z&sp=r&P1=1579201969&P2=11&P3=2&P4=lKDYcrhMcfeD75HvPn3Y19rU94ccBKYN83qNwaa6qHs%3d (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')],)",),))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\knack\knack\cli.py", line 206, in invoke
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\azure-cli-core\azure\cli\core\commands\__init__.py", line 603, in execute
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\azure-cli-core\azure\cli\core\commands\__init__.py", line 661, in _run_jobs_serially
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\azure-cli-core\azure\cli\core\commands\__init__.py", line 652, in _run_job
  File "C:\Users\ivaldez\.azure\cliextensions\azure-devops\azext_devops\dev\common\exception_handler.py", line 31, in azure_devops_exception_handler
    reraise(*sys.exc_info())
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\six\six.py", line 693, in reraise
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\azure-cli-core\azure\cli\core\commands\__init__.py", line 631, in _run_job
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\azure-cli-core\azure\cli\core\commands\__init__.py", line 306, in __call__
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\azure-cli-core\azure\cli\core\__init__.py", line 485, in default_command_handler
  File "C:\Users\ivaldez\.azure\cliextensions\azure-devops\azext_devops\dev\artifacts\universal.py", line 98, in download_package
    return artifact_tool.download_universal(organization, project, feed, name, version, path, file_filter)
  File "C:\Users\ivaldez\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool.py", line 43, in download_universal
    return self.run_artifacttool(organization, args, "Downloading")
  File "C:\Users\ivaldez\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool.py", line 58, in run_artifacttool
    artifacttool_dir = self._artifacttool_updater.get_latest_artifacttool(organization)
  File "C:\Users\ivaldez\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool_updater.py", line 42, in get_latest_artifacttool
    artifacttool_binary_path = self._get_artifacttool(organization)
  File "C:\Users\ivaldez\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool_updater.py", line 73, in _get_artifacttool
    _update_artifacttool(release_uri, release_id)
  File "C:\Users\ivaldez\.azure\cliextensions\azure-devops\azext_devops\dev\common\artifacttool_updater.py", line 96, in _update_artifacttool
    response = requests.get(uri, stream=True)
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\requests\requests\api.py", line 75, in get
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\requests\requests\api.py", line 60, in request
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\requests\requests\sessions.py", line 533, in request
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\requests\requests\sessions.py", line 646, in send
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-p8t_6awn\requests\requests\adapters.py", line 514, in send
requests.exceptions.SSLError: HTTPSConnectionPool(host='0t3vsblobprodcus362.vsblob.vsassets.io', port=443): Max retries exceeded with url: /artifacttool/artifacttool-win10-x64-Release_0.2.128.zip?sv=2017-04-17&sr=b&sig=EWrX6QfaHgakt2Wx2sORhYViPB7P0jNWWbeu3yGP3Dg%3D&spr=https&se=2020-01-16T19%3A17%3A49Z&sp=r&P1=1579201969&P2=11&P3=2&P4=lKDYcrhMcfeD75HvPn3Y19rU94ccBKYN83qNwaa6qHs%3d (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')],)",),))

To open an issue, please run: 'az feedback'`

**Comment by apluche** [comment]
@israelvaldez Could you provide more details on your corporate proxy settings and configuration?

As a side note, you may need to whitelist our storage account hosts/IPs on your proxy. Information is in the link below.
https://docs.microsoft.com/en-us/azure/devops/organizations/security/allow-list-ip-url?view=azure-devops

**Comment by rbpierce1** [comment]
As a note to any trying to solve a similar problem to the one above. We ran into the above issue, and isolated it to an SSL Inspection function in our network security application (ZScaler). Bypassing SSL Inspection for the URL solved the issue.

**Comment by eosfor** [comment]
@apluche @geverghe , colleagues, just stumbled upon this thread. We are looking to using universal packages as an "artifactory" for our deployments. There are cases where we cannot just use the tool, for instance we would like to be able to point azure webapps and their Kudu engine at an artifact via some HTTP URL, similar to a SAS token URL of a storage account. This would greatly simplify the process of deployment, and reduce the storage costs, as we wont need to store artifacts twice. 

So my questions are:

1. is there any way to download this programmatically, some c# class, or something?
2. is there a possibility to pull the package directly by HTTP URL, for example, if some rest call returns a SAS token, that points directly to the artifact. It could be a short-lived token, that is fine. The key thing here is that some target system should not authenticate to the URL, it just downloads the package directly from a URL.

Thanks,
Andrey

**Comment by samuelkoppes** [comment]
Closing as original issues are resolved. @eosfor please open a new issue with your questions.
